compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,212 @@
+use crate::transform::MirPass;
+use crate::util::patch::MirPatch;
+use rustc_data_structures::stable_map::FxHashMap;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Const, List, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_target::abi::{Size, Variants};
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
+
+impl<'tcx, const D: u64> MirPass<'tcx> for EnumSizeOpt<D> {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl<const D: u64> EnumSizeOpt<D> {
+    fn candidate<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        ty: Ty<'tcx>,
+        body_did: DefId,
+    ) -> Option<(Size, u64, Vec<Size>)> {
+        match ty.kind() {
+            ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                let p_e = tcx.param_env(body_did);
+                // FIXME(jknodt) handle error better below
+                let layout = tcx.layout_of(p_e.and(ty)).unwrap();
+                let variants = &layout.variants;
+                match variants {
+                    Variants::Single { .. } => None,
+                    Variants::Multiple { variants, .. } if variants.len() <= 1 => None,
+                    Variants::Multiple { variants, .. } => {
+                        let min = variants.iter().map(|v| v.size).min().unwrap();
+                        let max = variants.iter().map(|v| v.size).max().unwrap();
+                        if max.bytes() - min.bytes() < D {
+                            return None;
+                        }
+                        Some((
+                            layout.size,
+                            variants.len() as u64,
+                            variants.iter().map(|v| v.size).collect(),
```

Comment:
Hmm, I think that in a niche filling layout, a variant layout does not include the niche used for discriminant. Thus copying `variants[i].size` prefix of data could omit it and a determinant would be undefined.

Reply:
I see, for now I just left it super naive and made it so that all niche variants are ignored, don't know if that's an optimal solution or not.